crates/astria-conductor/src/data_availability.rs

```diff
@@ -0,0 +1,382 @@
+
+        let mut interval = tokio::time::interval(self.celestia_poll_interval);
+        loop {
+            select!(
+                shutdown_res = &mut self.shutdown => {
+                    match shutdown_res {
+                        Ok(()) => info!("received shutdown command; exiting"),
+                        Err(e) => warn!(error.message = %e, error.cause = ?e, "shutdown receiver dropped; exiting"),
+                    }
+                    break;
+                }
+
+                _ = interval.tick() => self.get_latest_height(),
+
+                res = async { self.get_latest_height.as_mut().unwrap().await }, if self.get_latest_height.is_some() => {
+                    self.get_latest_height = None;
+                    self.get_sequencer_datas(res);
+                }
+
+                Some((height, res)) = self.get_sequencer_datas.join_next(), if !self.get_sequencer_datas.is_empty() => {
+                    self.process_sequencer_datas(height, res);
+                }
+
+                Some((height, res)) = self.process_sequencer_datas.join_next(), if !self.process_sequencer_datas.is_empty() => {
+                    let span = tracing::info_span!("send_sequencer_subsets", height);
+                    span.in_scope(|| self.send_sequencer_subsets(res))
+                        .wrap_err("failed sending sequencer subsets to executor")?;
+                }
+            )
+        }
+        Ok(())
+    }
+
+    fn get_latest_height(&mut self) {
+        let client = self.celestia_client.inner().clone();
+        self.get_latest_height = Some(tokio::spawn(async move {
+            Ok(client.header_network_head().await?.height())
+        }))
+    }
+
+    fn get_sequencer_datas(&mut self, latest_height_res: Result<eyre::Result<u64>, JoinError>) {
+        let latest_height = match latest_height_res {
+            Err(e) => {
+                warn!(error.message = %e, error.cause = ?e, "task querying celestia for latest height failed");
+                return;
+            }
+
+            Ok(Err(e)) => {
+                warn!(error.message = %e, error.cause = ?e, "task querying celestia for latest height returned with an error");
+                return;
+            }
+
+            Ok(Ok(height)) => height,
+        };
+
+        if latest_height <= self.current_block_height {
+            info!(
+                height.celestia = latest_height,
+                height.previous = self.current_block_height,
+                "no new celestia height; not spawning tasks to fetch sequencer blocks"
+            );
+            return;
+        }
+        let first_new_height = self.current_block_height + 1;
+        info!(
+            height.start = first_new_height,
+            height.end = self.current_block_height,
+            "spawning tasks to fetch sequencer blocks for different celestia heights",
+        );
+        for height in first_new_height..=self.current_block_height {
```

Comment:
Shouldn't this be going to `latest_height` instead of `self.current_block_height`?

Reply:
if `let first_new_height = self.current_block_height + 1;` this doesn't execute?

Reply:
Yeah, that range would be considered empty.

Reply:
@joroshiba good catch, fixed.